aws/middleware/user_agent_test.go

```diff
@@ -0,0 +1,186 @@
+package middleware
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyhttp "github.com/awslabs/smithy-go/transport/http"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+var expectedAgent = aws.SDKName + "/" + aws.SDKVersion + " GOOS/" + runtime.GOOS + " GOARCH/" + runtime.GOARCH + " GO/" + runtime.Version()
+
+type mockBuilderHandler func(context.Context, middleware.BuildInput) (middleware.BuildOutput, middleware.Metadata, error)
+
+func (m mockBuilderHandler) HandleBuild(ctx context.Context, in middleware.BuildInput) (out middleware.BuildOutput, metadata middleware.Metadata, err error) {
+	return m(ctx, in)
+}
+
+func TestRequestUserAgent_HandleBuild(t *testing.T) {
+	cases := map[string]struct {
+		Env    map[string]string
+		In     middleware.BuildInput
+		Next   func(*testing.T, middleware.BuildInput) middleware.BuildHandler
+		Expect middleware.BuildInput
+		Err    bool
+	}{
+		"adds product information": {
+			In: middleware.BuildInput{Request: &smithyhttp.Request{
+				Request: &http.Request{Header: map[string][]string{}},
+			}},
+			Expect: middleware.BuildInput{Request: &smithyhttp.Request{
+				Request: &http.Request{Header: map[string][]string{
+					"User-Agent": {expectedAgent},
+				}},
+			}},
+			Next: func(t *testing.T, expect middleware.BuildInput) middleware.BuildHandler {
+				return mockBuilderHandler(func(ctx context.Context, input middleware.BuildInput) (o middleware.BuildOutput, m middleware.Metadata, err error) {
+					if diff := cmp.Diff(input, expect, cmpopts.IgnoreUnexported(http.Request{}, smithyhttp.Request{})); len(diff) > 0 {
+						t.Error(diff)
+					}
+					return o, m, err
+				})
```

Comment:
All of the `Next` member values look to be very similar if not the same. Would it make sense to move these all into a single implementation that can be referred for the tests cases? I think the `error for unknown type` case could still use the same handler.